File: packages/mobile_android/app/src/main/java/dev/verdeai/app/HostsModel.kt
package dev.verdeai.app

import androidx.lifecycle.ViewModel
import dev.verdeai.core.*
import kotlinx.coroutines.*
import kotlinx.coroutines.flow.*
import kotlinx.serialization.Serializable
import kotlinx.serialization.decodeFromString
import kotlinx.serialization.encodeToString
import java.util.UUID

@Serializable
internal data class SavedHost(val id: String, val label: String)
@Serializable
internal data class HostCatalog(val hosts: List<SavedHost>, val active: String?)
internal data class HostRow(val saved: SavedHost, val view: HostView? = null,
    val operation: Operation? = null, val busy: Boolean = false, val fatal: Boolean = false)
internal data class HostsState(val rows: List<HostRow> = emptyList(), val active: String? = null,
    val loading: Boolean = true, val busy: Boolean = false, val error: String? = null,
    val pairing: String? = null)

/** Owns platform host identities/selection only; every protocol action belongs to its core. */
internal class HostsModel(private val store: SecureStore,
    private val createHost: suspend (SavedHost) -> CoreHost) : ViewModel() {
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate)
    private val mutableState = MutableStateFlow(HostsState())
    val state = mutableState.asStateFlow()
    private val sessions = linkedMapOf<String, PairingModel>()
    private val observers = mutableMapOf<String, Job>()
    private val operations = mutableMapOf<String, String>()
    private var catalog = HostCatalog(emptyList(), null)
    private var foreground = false
    private var pendingLink: String? = null
    private var closed = false
    init { load() }

    fun pairing(id: String): PairingModel? = sessions[id]
    // D-05 consumes projections from this handle; runtime data is never merged across hosts.
    suspend fun activeCore(): CoreHost? = state.value.active?.let { sessions[it]?.coreHost() }

    fun load() {
        if (state.value.busy || closed) return
        mutableState.update { it.copy(busy=true, error=null) }
        scope.launch {
            try {
                val saved = store.get(CATALOG_KEY)
                val value = saved?.let { CoreJson.decodeFromString<HostCatalog>(it) }
                    ?: HostCatalog(listOf(SavedHost("primary", "My host")), "primary")
                require(value.hosts.size <= 32 && value.hosts.map { it.id }.distinct().size == value.hosts.size &&
                    value.hosts.all { it.id.matches(Regex("[A-Za-z0-9_-]{1,128}")) && it.label.isNotBlank() } &&
                    (value.active == null || value.hosts.any { it.id == value.active }))
                if (saved == null) persist(value) else catalog=value
                mutableState.update { it.copy(rows=value.hosts.map(::HostRow), active=value.active, loading=false) }
                value.hosts.forEach(::open)
            } catch (_: Exception) {
                mutableState.update { it.copy(error="Could not load saved hosts. Unlock your phone and retry.") }
            } finally {
                mutableState.update { it.copy(busy=false) }
                if (!state.value.loading) pendingLink?.let { pendingLink=null; receiveLink(it) }
            }
        }
    }
    private fun open(saved: SavedHost) {
        if (sessions.containsKey(saved.id)) return
        val pairing = PairingModel { createHost(saved) }
        sessions[saved.id] = pairing
        pairing.foreground(foreground)
        observers[saved.id] = scope.launch {
            launch { pairing.state.collect { value ->
                row(saved.id) { it.copy(view=value.host, fatal=value.fatal) }
            } }
            try { pairing.coreHost().hosts.collect { value ->
                row(saved.id) { it.copy(operation=value?.data?.operations?.find { op -> op.intent_id == operations[saved.id] }) }
            } } catch (_: Exception) { row(saved.id) { it.copy(fatal=true) } }
        }
    }
    private fun row(id: String, change: (HostRow) -> HostRow) {
        mutableState.update { it.copy(rows=it.rows.map { row -> if (row.saved.id == id) change(row) else row }) }
    }
    private suspend fun persist(value: HostCatalog) {
        store.put(CATALOG_KEY, CoreJson.encodeToString(value))
        catalog=value
    }
    private fun catalogAction(block: suspend () -> Unit) {
        if (state.value.busy || state.value.loading || closed) return
        mutableState.update { it.copy(busy=true, error=null) }
        scope.launch {
            try { block() }
            catch (_: Exception) { mutableState.update { it.copy(error="Could not save hosts. Unlock your phone and retry.") } }
            finally { mutableState.update { it.copy(busy=false) } }
        }
    }
    fun select(id: String) = catalogAction {
        if (catalog.hosts.none { it.id == id }) return@catalogAction
        persist(catalog.copy(active=id))
        mutableState.update { it.copy(active=id, pairing=null) }
    }
    fun add(label: String, link: String? = null) = catalogAction {
        if (catalog.hosts.size >= 32) {
            mutableState.update { it.copy(error="Remove a host before adding another (32 host limit).") }
            return@catalogAction
        }
        val saved = SavedHost(UUID.randomUUID().toString(), label.trim().take(128).ifEmpty { "My host" })
        persist(HostCatalog(catalog.hosts + saved, saved.id))
        mutableState.update { it.copy(rows=it.rows + HostRow(saved), active=saved.id, pairing=saved.id) }
        open(saved)
        link?.let { sessions[saved.id]?.receiveLink(it) }
    }
    fun showPairing(id: String?) { mutableState.update { it.copy(pairing=id) } }
    fun receiveLink(link: String) {
        if (state.value.loading) { pendingLink=link; return }
        val id = state.value.pairing ?: state.value.active
        val model = sessions[id]
        if (model != null && !model.state.value.complete) {
            model.receiveLink(link)
            showPairing(id)
        } else add("My host", link)
    }
    fun foreground(active: Boolean) {
        foreground=active
        sessions.values.forEach { it.foreground(active) }
    }
    fun signOut(id: String, forget: Boolean = false) = action(id) { core, intent ->
        core.send { n,w -> if (forget) EventForgetHost(now_ms=n, wall_time_ms=w, intent_id=intent, host_id=id)
            else EventSignOut(now_ms=n, wall_time_ms=w, intent_id=intent, host_id=id) }
    }
    fun retry(id: String) = action(id) { core, intent ->
        core.send { n,w -> EventRetryConnection(now_ms=n, wall_time_ms=w, intent_id=intent) }
    }
    private fun action(id: String, block: suspend (CoreHost, String) -> Unit) {
        val row = state.value.rows.find { it.saved.id == id } ?: return
        if (row.busy || row.operation?.state == "pending" || row.fatal) return
        val session = sessions[id] ?: return
        val intent = UUID.randomUUID().toString()
        operations[id]=intent
        row(id) { it.copy(busy=true, operation=null) }
        scope.launch {
            try { block(session.coreHost(), intent) }
            catch (_: CoreInputRejected) { mutableState.update { it.copy(error="This host is still loading or finishing an action. Try again shortly.") } }
            catch (_: Exception) { row(id) { it.copy(fatal=true) } }
            finally { row(id) { it.copy(busy=false) } }
        }
    }
    /** Catalog removal is allowed only after the core confirms its secure-store deletes. */
    fun remove(id: String) = catalogAction {
        if (state.value.rows.find { it.saved.id == id }?.view?.auth_state != "signed_out") return@catalogAction
        val remaining = catalog.hosts.filterNot { it.id == id }
        val active = if (catalog.active == id) remaining.firstOrNull()?.id else catalog.active
        persist(HostCatalog(remaining, active))
        observers.remove(id)?.cancel()
        sessions.remove(id)?.dispose()
        operations.remove(id)
        mutableState.update { it.copy(rows=it.rows.filterNot { row -> row.saved.id == id }, active=active,
            pairing=it.pairing?.takeUnless { pair -> pair == id }) }
    }
    override fun onCleared() {
        closed=true
        pendingLink=null
        sessions.values.forEach { it.dispose() }
        scope.cancel()
    }
    companion object { const val CATALOG_KEY = "android/1/hosts" }
}

internal fun hostStatus(row: HostRow): String = when {
    row.fatal -> "Connection unavailable — reopen Verde"
    row.view == null || row.view.auth_state == "loading" -> "Loading"
    row.view.auth_state == "signed_out" -> "Signed out"
    row.view.auth_state == "signing_out" -> "Removing local data"
    row.view.auth_state == "repair_required" -> "Pair again — device authorization needs renewal"
    row.view.trust_proposal != null -> "Review host identity"
    row.view.update_required -> "Update required"
    row.view.auth_state == "unpaired" -> "Not paired"
    row.view.phase == "ready" -> "Connected"
    row.view.error?.failure_kind == "network" || row.view.phase == "failed" -> "Unreachable — is Tailscale on?"
    row.view.lifecycle == Lifecycle.background || row.view.phase == "disabled" -> "Offline"
    else -> "Connecting"
}

File: packages/mobile_ios/App/HostsModel.swift
import Foundation
import Observation

struct SavedHost: Codable, Equatable {
    var id: String
    var label: String
}

struct HostCatalog: Codable, Equatable {
    var hosts: [SavedHost]
    var active: String?
}

/// Unambiguous name for the generated core model (Foundation also defines `Operation`).
typealias CoreOperation = Operation

struct HostRow {
    var saved: SavedHost
    var view: HostView?
    var operation: Operation?
    var busy = false
    var fatal = false
}

/// Owns platform host identities and selection only. Every protocol action
/// (pairing, sign-out, forget, retry) belongs to that host's own core handle,
/// and runtime data is never merged across hosts.
@MainActor @Observable
final class HostsModel {
    static let catalogKey = "ios/1/hosts"
    static let maxHosts = 32
    /// Auth states whose local data is gone or going; the warm-start cache follows.
    static let wiped: Set<String> = ["unpaired", "signing_out", "signed_out"]
    /// Written by the I-03 single-host build; migrated into the catalog once.
    static let legacyDefaultsKey = "pairing.hostID"

    private(set) var catalog = HostCatalog(hosts: [], active: nil)
    private(set) var loading = true
    private(set) var busy = false
    var error: String?
    private(set) var pairing: String?
    private(set) var sessions: [String: PairingModel] = [:]
    private var intents: [String: String] = [:]
    private var acting: Set<String> = []
    private var cleared: Set<String> = []
    private(set) var isForeground = false
    private(set) var networkState: NetworkState?
    private var pendingLink: String?
    private var began = false
    private var closed = false
    /// Browse/cache observers: the active host changed, or a host's store published.
    @ObservationIgnored var onActiveChange: (() -> Void)?
    @ObservationIgnored var onStoreChange: ((String) -> Void)?

    private let storage: SecureStorage
    let cache: ViewCache?
    private let deviceLabel: String
    private let legacyID: () -> String?
    private let makeHost: (SavedHost, CoreViewStore) throws -> CoreHost

    init(storage: SecureStorage, cache: ViewCache?, deviceLabel: String,
         legacyID: @escaping () -> String? = { nil },
         makeHost: @escaping (SavedHost, CoreViewStore) throws -> CoreHost) {
        self.storage = storage
        self.cache = cache
        self.deviceLabel = deviceLabel
        self.legacyID = legacyID
        self.makeHost = makeHost
    }

    static func live(deviceLabel: String) -> HostsModel {
        HostsModel(storage: KeychainStorage(), cache: ViewCache.live(), deviceLabel: deviceLabel,
                   legacyID: { UserDefaults.standard.string(forKey: legacyDefaultsKey) }) { saved, store in
            try CoreHost.live(hostID: saved.id, label: saved.label, httpsURL: nil, wssURL: nil, store: store)
        }
    }

    static func validID(_ id: String) -> Bool {
        (1...128).contains(id.utf8.count) && id.utf8.allSatisfy {
            ($0 >= 0x30 && $0 <= 0x39) || ($0 >= 0x41 && $0 <= 0x5a) || ($0 >= 0x61 && $0 <= 0x7a) || $0 == 0x2d || $0 == 0x5f
        }
    }

    static func valid(_ value: HostCatalog) -> Bool {
        let ids = value.hosts.map(\.id)
        return value.hosts.count <= maxHosts && Set(ids).count == ids.count
            && value.hosts.allSatisfy { validID($0.id) && !$0.label.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty && $0.label.count <= 128 }
            && (value.active.map { ids.contains($0) } ?? true)
    }

    // MARK: Rows

    var active: String? { catalog.active }
    var rows: [HostRow] { catalog.hosts.map { makeRow($0) } }
    func row(_ id: String) -> HostRow? { catalog.hosts.first { $0.id == id }.map { makeRow($0) } }
    func session(_ id: String?) -> PairingModel? { id.flatMap { sessions[$0] } }

    private func makeRow(_ saved: SavedHost) -> HostRow {
        let session = sessions[saved.id]
        let intent = intents[saved.id]
        return HostRow(saved: saved, view: session?.row,
                       operation: intent.flatMap { id in session?.store.hosts?.data?.operations.first { $0.intent_id == id } },
                       busy: acting.contains(saved.id), fatal: session?.fatal ?? false)
    }

    // MARK: Catalog

    /// First load, once the app has its initial signals; later calls are no-ops.
    func begin() {
        guard !began else { return }
        began = true
        load()
    }

    func load() {
        guard loading, !closed else { return }
        busy = true
        error = nil
        do {
            let saved = try storage.get(Self.catalogKey)
            let value: HostCatalog
            if let saved {
                value = try JSONDecoder().decode(HostCatalog.self, from: saved)
            } else {
                // First launch after I-03 keeps that host's Keychain records; a fresh install gets one slot.
                let id = legacyID().flatMap { Self.validID($0) ? $0 : nil } ?? "primary"
                value = HostCatalog(hosts: [SavedHost(id: id, label: "My host")], active: id)
            }
            guard Self.valid(value) else { throw StorageError(code: .io) }
            if saved == nil { try persist(value) } else { setCatalog(value) }
            loading = false
            value.hosts.forEach { open($0) }
        } catch {
            self.error = "Could not load saved hosts. Unlock your phone and retry."
        }
        busy = false
        if !loading, let link = pendingLink { pendingLink = nil; receiveLink(link) }
    }

    private func persist(_ value: HostCatalog) throws {
        try storage.put(Self.catalogKey, value: JSONEncoder().encode(value))
        setCatalog(value)
    }

    private func setCatalog(_ value: HostCatalog) {
        let changed = value.active != catalog.active
        catalog = value
        if changed { onActiveChange?() }
    }

    private func catalogAction(_ block: () throws -> Void) {
        guard !busy, !loading, !closed else { return }
        error = nil
        do { try block() } catch { self.error = "Could not save hosts. Unlock your phone and retry." }
    }

    func select(_ id: String) {
        catalogAction {
            guard catalog.hosts.contains(where: { $0.id == id }) else { return }
            try persist(HostCatalog(hosts: catalog.hosts, active: id))
            pairing = nil
        }
    }

    func add(label: String, link: String? = nil) {
        catalogAction {
            guard catalog.hosts.count < Self.maxHosts else {
                error = "Remove a host before adding another (32 host limit)."
                return
            }
            let trimmed = String(label.trimmingCharacters(in: .whitespacesAndNewlines).prefix(128))
            let saved = SavedHost(id: UUID().uuidString.lowercased(), label: trimmed.isEmpty ? "My host" : trimmed)
            try persist(HostCatalog(hosts: catalog.hosts + [saved], active: saved.id))
            pairing = saved.id
            open(saved)
            if let link, let session = sessions[saved.id] { Task { await session.receive(link) } }
        }
    }

    /// Catalog removal is allowed only after the core confirms its secure-store deletes.
    func remove(_ id: String) {
        catalogAction {
            guard row(id)?.view?.auth_state == "signed_out" else { return }
            let remaining = catalog.hosts.filter { $0.id != id }
            let active = catalog.active == id ? remaining.first?.id : catalog.active
            try persist(HostCatalog(hosts: remaining, active: active))
            if let session = sessions.removeValue(forKey: id) { Task { await session.stop() } }
            intents[id] = nil
            acting.remove(id)
            cleared.remove(id)
            cache?.clear(id)
            if pairing == id { pairing = nil }
        }
    }

    func showPairing(_ id: String?) { pairing = id }

    // MARK: Sessions and signals

    private func open(_ saved: SavedHost) {
        guard sessions[saved.id] == nil else { return }
        let store = CoreViewStore()
        let id = saved.id
        let makeHost = self.makeHost
        let session = PairingModel(deviceLabel: deviceLabel, store: store) { store in try makeHost(saved, store) }
        store.onApply = { [weak self] in self?.storeChanged(id) }
        sessions[id] = session
        // Queued behind `start`: network first, then the current foreground state.
        if let networkState { session.network(networkState) }
        session.foreground(isForeground)
        Task { await session.start() }
    }

    private func storeChanged(_ id: String) {
        guard let session = sessions[id] else { return }
        if let state = session.row?.auth_state {
            // Local removal or a missing credential retires this host's warm-start cache.
            if Self.wiped.contains(state) {
                if !cleared.contains(id) { cleared.insert(id); cache?.clear(id) }
            } else if state == "paired" { cleared.remove(id) }
        }
        if session.awaitingSubmit { Task { await session.submitPending() } }
        onStoreChange?(id)
    }

    /// Pairing links (scanned, pasted, universal or custom-scheme) go to the host
    /// being paired, else to a new host slot.
    func receiveLink(_ link: String) {
        guard !closed else { return }
        if loading { pendingLink = link; return }
        let id = pairing ?? catalog.active
        if let id, let session = sessions[id], !session.complete {
            pairing = id
            Task { await session.receive(link) }
        } else {
            add(label: "My host", link: link)
        }
    }

    func open(url: URL) {
        guard PairingInput.routes(url) else { return }
        receiveLink(url.absoluteString)
    }

    func foreground(_ active: Bool) {
        isForeground = active
        sessions.values.forEach { $0.foreground(active) }
    }

    func network(_ state: NetworkState) {
        networkState = state
        sessions.values.forEach { $0.network(state) }
    }

    // MARK: Host actions

    func signOut(_ id: String, forget: Bool = false) {
        action(id) { host, intent in
            try await host.send(forget
                ? .forget_host(EventForgetHost(now_ms: 0, wall_time_ms: 0, intent_id: intent, host_id: id))
                : .sign_out(EventSignOut(now_ms: 0, wall_time_ms: 0, intent_id: intent, host_id: id)))
        }
    }

    /// Also retries a failed local delete: the core resumes deletion without repeating revoke.
    func retry(_ id: String) {
        action(id) { host, intent in
            try await host.send(.retry_connection(EventRetryConnection(now_ms: 0, wall_time_ms: 0, intent_id: intent)))
        }
    }

    private func action(_ id: String, _ block: @escaping (CoreHost, String) async throws -> Void) {
        guard !closed, let row = row(id), !row.busy, row.operation?.state != "pending", !row.fatal,
              let session = sessions[id] else { return }
        let intent = UUID().uuidString
        intents[id] = intent
        acting.insert(id)
        Task {
            defer { acting.remove(id) }
            await session.start()
            guard let host = session.host else { return } // Open failure is already published as fatal.
            do { try await block(host, intent) }
            catch CoreBridgeError.rejected { error = "This host is still loading or finishing an action. Try again shortly." }
            catch { /* The host closed and its store published a fixed failure. */ }
        }
    }

    func close() async {
        closed = true
        pendingLink = nil
        let all = Array(sessions.values)
        sessions.removeAll()
        for session in all { await session.stop() }
    }
}

func hostStatus(_ row: HostRow) -> String {
    if row.fatal { return "Connection unavailable — reopen Verde" }
    guard let view = row.view, view.auth_state != "loading" else { return "Loading" }
    if view.auth_state == "signed_out" { return "Signed out" }
    if view.auth_state == "signing_out" { return "Removing local data" }
    if view.auth_state == "repair_required" { return "Pair again — device authorization needs renewal" }
    if view.trust_proposal != nil { return "Review host identity" }
    if view.update_required { return "Update required" }
    if view.auth_state == "unpaired" { return "Not paired" }
    if view.phase == "ready" { return "Connected" }
    if view.error?.failure_kind == "network" || view.phase == "failed" { return "Unreachable — is Tailscale on?" }
    if view.lifecycle == .background || view.phase == "disabled" { return "Offline" }
    return "Connecting"
}
